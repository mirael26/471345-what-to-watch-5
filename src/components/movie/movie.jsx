import React, {PureComponent} from "react";
import {connect} from "react-redux";
import PropTypes from "prop-types";
import {Link} from "react-router-dom";

import {fetchCurrentFilm} from "../../store/api-actions";
import {AuthorizationStatus, AppRoute} from "../../const";

import UserBlock from "../user-block/user-block";
import MoviesList from "../movies-list/movies-list";
import MovieInfo from "../movie-info/movie-info";
import withViewType from "../../hocs/with-view-type/with-view-type";

const MAX_SIMILAR_FILMS = 4;

const MovieInfoWithViewType = withViewType(MovieInfo);

class Movie extends PureComponent {
  constructor(props) {
    super(props);
    this.getFilmId = this.getFilmId.bind(this);
  }

  componentDidMount() {
    const {loadCurrentFilm} = this.props;
    loadCurrentFilm(this.getFilmId());
  }

  componentDidUpdate(prevProps) {
    if (this.props.match.params.id !== prevProps.match.params.id) {
      this.props.loadCurrentFilm(this.getFilmId());
    }
  }

  getFilmId() {
    const {match} = this.props;
    const {params: {id}} = match;
    return parseInt(id, 10);
  }

  render() {
    const {
      films,
      currentFilm,
      authorizationStatus,
    } = this.props;
    const filmId = this.getFilmId();

    const similarFilms = films
      .filter((film) => {
        return (film.genre === currentFilm.genre) && (film.id !== currentFilm.id);
      })
      .slice(0, MAX_SIMILAR_FILMS);
    const isAuthorized = authorizationStatus === AuthorizationStatus.AUTH;

    return Object.keys(currentFilm).length === 0 ? null : (<React.Fragment>
      <section className="movie-card movie-card--full" style={{backgroundColor: currentFilm.backgroundColor}}>
        <div className="movie-card__hero">
          <div className="movie-card__bg">
            <img src={currentFilm.backgroundPicture} alt={currentFilm.title} />
          </div>

          <h1 className="visually-hidden">WTW</h1>

          <header className="page-header movie-card__head">
            <div className="logo">
              <Link to={AppRoute.MAIN} className="logo__link">
                <span className="logo__letter logo__letter--1">W</span>
                <span className="logo__letter logo__letter--2">T</span>
                <span className="logo__letter logo__letter--3">W</span>
              </Link>
            </div>

            {<UserBlock />}
          </header>

          <div className="movie-card__wrap">
            <div className="movie-card__desc">
              <h2 className="movie-card__title">{currentFilm.title}</h2>
              <p className="movie-card__meta">
                <span className="movie-card__genre">{currentFilm.genre}</span>
                <span className="movie-card__year">{currentFilm.releaseDate}</span>
              </p>

              <div className="movie-card__buttons">
                <Link to={`${AppRoute.PLAYER}/${filmId}`} className="btn btn--play movie-card__button" type="button">
                  <svg viewBox="0 0 19 19" width="19" height="19">
                    <use xlinkHref="#play-s"></use>
                  </svg>
                  <span>Play</span>
                </Link>
                <button className="btn btn--list movie-card__button" type="button">
                  <svg viewBox="0 0 19 20" width="19" height="20">
                    <use xlinkHref="#add"></use>
                  </svg>
                  <span>My list</span>
                </button>
                {isAuthorized ? <Link to={`${AppRoute.MOVIE}/${filmId}${AppRoute.REVIEW}`} className="btn movie-card__button">Add review</Link> : ``}
              </div>
            </div>
          </div>
        </div>

        <div className="movie-card__wrap movie-card__translate-top">
          <div className="movie-card__info">
            <div className="movie-card__poster movie-card__poster--big">
              <img src={currentFilm.poster} alt={currentFilm.title} width="218" height="327" />
            </div>

            <MovieInfoWithViewType
              film={currentFilm}
              id={filmId}
            />
          </div>
        </div>
      </section>

      <div className="page-content">
        <section className="catalog catalog--like-this">
          <h2 className="catalog__title">More like this</h2>

          {<MoviesList films={similarFilms}/>}
        </section>

        <footer className="page-footer">
          <div className="logo">
            <Link to={AppRoute.MAIN} className="logo__link logo__link--light">
              <span className="logo__letter logo__letter--1">W</span>
              <span className="logo__letter logo__letter--2">T</span>
              <span className="logo__letter logo__letter--3">W</span>
            </Link>
          </div>

          <div className="copyright">
            <p>Â© 2019 What to watch Ltd.</p>
          </div>
        </footer>
      </div>
    </React.Fragment>);
  }
}

Movie.propTypes = {
  films: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    genre: PropTypes.string.isRequired,
    releaseDate: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    description: PropTypes.string.isRequired,
    picture: PropTypes.string.isRequired,
    poster: PropTypes.string.isRequired,
    backgroundPicture: PropTypes.string.isRequired,
    backgroundColor: PropTypes.string,
    rate: PropTypes.number.isRequired,
    voteCount: PropTypes.number.isRequired,
    director: PropTypes.string.isRequired,
    starring: PropTypes.arrayOf(PropTypes.string).isRequired,
    runTime: PropTypes.number.isRequired,
    previewVideoSrc: PropTypes.string.isRequired,
  })).isRequired,
  currentFilm: PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
    genre: PropTypes.string,
    releaseDate: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    description: PropTypes.string,
    picture: PropTypes.string,
    poster: PropTypes.string,
    backgroundPicture: PropTypes.string,
    backgroundColor: PropTypes.string,
    rate: PropTypes.number,
    voteCount: PropTypes.number,
    director: PropTypes.string,
    starring: PropTypes.arrayOf(PropTypes.string),
    runTime: PropTypes.number,
    previewVideoSrc: PropTypes.string,
  }),
  match: PropTypes.object.isRequired,
  loadCurrentFilm: PropTypes.func.isRequired,
  authorizationStatus: PropTypes.string.isRequired,
};

const mapStateToProps = ({DATA, USER}) => ({
  films: DATA.films,
  currentFilm: DATA.currentFilm,
  authorizationStatus: USER.authorizationStatus,
});

const mapDispatchToProps = (dispatch) => ({
  loadCurrentFilm(id) {
    dispatch(fetchCurrentFilm(id));
  }
});

export {Movie};
export default connect(mapStateToProps, mapDispatchToProps)(Movie);
