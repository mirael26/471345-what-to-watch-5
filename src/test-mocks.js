const films = [
  {
    id: 1,
    title: `The Grand Budapest Hotel`,
    genre: `Drama`,
    releaseDate: `2014`,
    description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.

    Gustave prides himself on providing first-class service to the hotel's guests, including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief suspect in her murder.`,
    picture: `https://www.theglobeandmail.com/resizer/ZlZ4o9jXtk_x8e20WPaKO1sFD6I=/1200x0/filters:quality(80)/arc-anglerfish-tgam-prod-tgam.s3.amazonaws.com/public/CEK54CW3HZBMXKO3QNV6PHHNEE`,
    poster: `img/the-grand-budapest-hotel-poster.jpg`,
    backgroundPicture: `img/bg-the-grand-budapest-hotel.jpg`,
    backgroundColor: `ffffff`,
    rate: 8.9,
    voteCount: 240,
    director: `Wes Andreson`,
    starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`, `Saoirse Ronan`, `Tony Revoloru`, `Tilda Swinton`, `Tom Wilkinson`, `Owen Wilkinson`, `Adrien Brody`, `Ralph Fiennes`, `Jeff Goldblum`],
    runTime: 99,
    previewVideoSrc: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    videoSrc: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    isFavorite: true,
  },
  {
    description: `The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence ,and redemption.`,
    director: `Quentin Tarantino`,
    genre: `Crime`,
    id: 2,
    title: `Pulp Fiction`,
    releaseDate: 1994,
    picture: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/pulp-fiction.jpg`,
    poster: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg`,
    backgroundPicture: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg`,
    backgroundColor: `#795433`,
    starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`, `Saoirse Ronan`, `Tony Revoloru`, `Tilda Swinton`, `Tom Wilkinson`, `Owen Wilkinson`, `Adrien Brody`, `Ralph Fiennes`, `Jeff Goldblum`],
    rate: 1.5,
    voteCount: 1635992,
    runTime: 153,
    videoSrc: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    previewVideoSrc: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.,ogv.360p.webm`,
    isFavorite: false,
  },
  {
    description: `Unscrupulous boxing promoters, violent bookmakers, a Russian gangster, incompetent amateur robbers and supposedly Jewish jewelers ,fight to track down a priceless stolen diamond.`,
    director: `Guy Ritchie`,
    genre: `Comedy`,
    id: 13,
    title: `Snatch`,
    releaseDate: 2000,
    picture: `https: //assets.htmlacademy.ru/intensives/javascript-3/film/preview/snatch.jpg`,
    poster: `https: //assets.htmlacademy.ru/intensives/javascript-3/film/poster/Snatch.jpg`,
    backgroundPicture: `https: //assets.htmlacademy.ru/intensives/javascript-3/film/background/Snatch.jpg`,
    backgroundColor: `#FDFDFC`,
    starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`, `Saoirse Ronan`, `Tony Revoloru`, `Tilda Swinton`, `Tom Wilkinson`, `Owen Wilkinson`, `Adrien Brody`, `Ralph Fiennes`, `Jeff Goldblum`],
    rate: 0.2,
    voteCount: 716577,
    runTime: 104,
    videoSrc: `http: //media.xiph.org/mango/tears_of_steel_1080p.webm`,
    previewVideoSrc: `https: //upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.,360p.webm`,
    isFavorite: false,
  },
];

const reviews = [
  {
    id: 1,
    text: `Discerning travellers and Wes Anderson fans will luxuriate in the glorious Mittel-European kitsch of one of the director's funniest and most exquisitely designed movies in years.`,
    rate: 8.9,
    author: `Kate Muir`,
    date: `2016-12-24`,
  },
  {
    id: 2,
    text: `Anderson's films are too precious for some, but for those of us willing to lose ourselves in them, they're a delight. "The Grand Budapest Hotel" is no different, except that he has added a hint of gravitas to the mix, improving the recipe.`,
    rate: 8.0,
    author: `Bill Goodykoontz`,
    date: `2015-11-18`,
  },
  {
    id: 3,
    text: `I didn't find it amusing, and while I can appreciate the creativity, it's an hour and 40 minutes I wish I could take back.`,
    rate: 8.0,
    author: `Amanda Greever`,
    date: `2015-11-18`,
  },
  {
    id: 4,
    text: `The mannered, madcap proceedings are often delightful, occasionally silly, and here and there, gruesome and/or heartbreaking.`,
    rate: 7.2,
    author: `Matthew Lickona`,
    date: `2016-12-20`,
  },
];

const genres = [
  `Action`,
  `Crime`,
  `Drama`,
  `Adventure`,
  `Comedy`,
  `Fantasy`,
  `Thriller`,
];

const match = {
  params: {
    id: 1,
  }
};

export {films, reviews, genres, match};
