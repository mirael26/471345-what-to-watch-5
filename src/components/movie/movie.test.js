import React from "react";
import renderer from "react-test-renderer";
import {Router} from "react-router-dom";
import browserHistory from "../../browser-history";

import {films, match} from "../../test-mocks";

import {Movie} from "./movie";

jest.mock(`../user-block/user-block`, () => `UserBlock`);
jest.mock(`../movies-list/movies-list`, () => `MoviesList`);
jest.mock(`../movie-info/movie-info`, () => `MovieInfo`);

it(`Should Movie render correctly`, () => {
  const noop = () => {};
  const tree = renderer
    .create(
        <Router history={browserHistory}>
          <Movie
            films={films}
            currentFilm={films[1]}
            match={match}
            loadCurrentFilm={noop}
            authorizationStatus={`AUTH`}
            onPlayerButtonClick={noop}
            changeFavoriteStatus={noop}
            onUnauthorizedFavoriteClick={noop}
            loadingErrorStatus={false}
          />
        </Router>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
