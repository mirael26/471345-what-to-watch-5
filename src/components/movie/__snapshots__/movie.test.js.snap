// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Movie render correctly 1`] = `
Array [
  <section
    className="movie-card movie-card--full"
    style={
      Object {
        "backgroundColor": "#795433",
      }
    }
  >
    <div
      className="movie-card__hero"
    >
      <div
        className="movie-card__bg"
      >
        <img
          alt="Pulp Fiction"
          src="https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg"
        />
      </div>
      <h1
        className="visually-hidden"
      >
        WTW
      </h1>
      <header
        className="page-header movie-card__head"
      >
        <div
          className="logo"
        >
          <a
            className="logo__link"
            href="/"
            onClick={[Function]}
          >
            <span
              className="logo__letter logo__letter--1"
            >
              W
            </span>
            <span
              className="logo__letter logo__letter--2"
            >
              T
            </span>
            <span
              className="logo__letter logo__letter--3"
            >
              W
            </span>
          </a>
        </div>
        <UserBlock />
      </header>
      <div
        className="movie-card__wrap"
      >
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            Pulp Fiction
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Crime
            </span>
            <span
              className="movie-card__year"
            >
              1994
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              className="btn btn--list movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlinkHref="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
            <a
              className="btn movie-card__button"
              href="/films/1/review"
              onClick={[Function]}
            >
              Add review
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      className="movie-card__wrap movie-card__translate-top"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster movie-card__poster--big"
        >
          <img
            alt="Pulp Fiction"
            height="327"
            src="https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg"
            width="218"
          />
        </div>
        <MovieInfo
          film={
            Object {
              "backgroundColor": "#795433",
              "backgroundPicture": "https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg",
              "description": "The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence ,and redemption.",
              "director": "Quentin Tarantino",
              "genre": "Crime",
              "id": 2,
              "isFavorite": false,
              "picture": "https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/pulp-fiction.jpg",
              "poster": "https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg",
              "previewVideoSrc": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.,ogv.360p.webm",
              "rate": 1.5,
              "releaseDate": 1994,
              "runTime": 153,
              "starring": Array [
                "Bill Murray",
                "Edward Norton",
                "Jude Law",
                "Willem Dafoe",
                "Saoirse Ronan",
                "Tony Revoloru",
                "Tilda Swinton",
                "Tom Wilkinson",
                "Owen Wilkinson",
                "Adrien Brody",
                "Ralph Fiennes",
                "Jeff Goldblum",
              ],
              "title": "Pulp Fiction",
              "videoSrc": "http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4",
              "voteCount": 1635992,
            }
          }
          id={1}
          onTabClick={[Function]}
          viewType="overview"
        />
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog catalog--like-this"
    >
      <h2
        className="catalog__title"
      >
        More like this
      </h2>
      <MoviesList
        films={Array []}
      />
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
